systems/introduction/vm.go

```diff
@@ -28,16 +30,47 @@ const (
 func compute(memory []byte) {
 
 	registers := [3]byte{8, 0, 0} // PC, R1 and R2
+	var pcincr byte               // Amount to increment the PC each cycle
 
 	// Keep looping, like a physical computer's clock
 	for {
+		pcincr = 3 // Default instruction length is three bytes
+
+		op := memory[registers[0]]
+		arg1 := memory[registers[0]+1]
+		arg2 := memory[registers[0]+2]
 
-		// op := TODO // fetch the opcode
+		// decode and execute
+		switch op {
+		case Load:
+			registers[arg1] = memory[arg2]
+		case Store:
+			if arg2 >= 8 {
+				// We don't want to write in the instructions portion of memory
+				panic(fmt.Errorf("Segmentation fault: Can't write to location 0x%x", arg2))
```

Comment:
I'm not a Go expert, but one of the pieces of feedback I got in my Go pre-work was not using `panic` for things other than bugs. Rather, returning an error to the caller is often preferred. You'll see this in the Go book on page 128.

So perhaps adding a return value that's an error to the `compute` function is better? Then you can return `nil` if the compute function succeeds.

I'd like to get @rouxcaesar 's take on this since he's way better at Go than I am.

Reply:
Ah, that's a good callout. Honestly I wasn't sure whether to `panic` or not in this case so I appreciate the feedback.